torch/fx/experimental/symbolic_shapes.py

```diff
@@ -186,79 +177,225 @@ def magic_impl(self, other):
+        """
+        acopy = a
+        while a in self.replacements:
+            a = self.replacements[a]
+        while acopy in self.replacements:
+            self.replacements[acopy], acopy = a, self.replacements[acopy]
+        return a
+
+    def _maybe_guard_eq(self, expr: "sympy.Eq") -> None:
+        """
+        Evaluates the result of an eq call. If true, uses information to
+        simplify shapes (i.e. a == b or a % 5 == 0)
+        """
+        concrete_bool = bool(self.size_hint(expr))
+        if not concrete_bool:
+            return
+        free = list(expr.free_symbols)
+
+        assert len(free) > 0, "The expression should not be static by this point"
+        # In case of really gnarly expression, we don't blow up
+        if len(free) <= 4:
+            free = sorted(free, key=lambda x: (self.size_hint(x), x.name), reverse=True)  # type: ignore[attr-defined]
+            lhs = expr.lhs
+            rhs = expr.rhs
+            try:
+                solutions = sympy.solveset(lhs - rhs, free[0], domain=sympy.S.Integers)
+                if not solutions.is_finite_set:
+                    if len(expr.atoms(sympy.Mod)) == 1:
+                        mod_expr = tuple(expr.atoms(sympy.Mod))[0]
+                        solutions = sympy.solveset(lhs - rhs, mod_expr, domain=sympy.S.Integers)
+                        if solutions.is_finite_set and len(solutions) == 1 and tuple(solutions)[0] == 0:
+                            self.divisible[mod_expr] = 0
+                    return
+
+                if not isinstance(solutions, sympy.FiniteSet):
+                    return
+
+                solutions = tuple(solutions)
+                if len(solutions) == 1 and "/" not in str(solutions[0]):
+                    new_var = solutions[0]
+                    new_var = self._find(solutions[0])
+                    self.replacements[free[0]] = new_var
+            except ZeroDivisionError:
+                pass
+
+        return
+
+    @lru_cache(256)
+    def evaluate_expr(self, expr: "sympy.Expr"):
+        """
+        Given an expression, evaluates it, adding guards if necessary
+        """
+        try:
+            if len(list(expr.free_symbols)) == 0:
+                return expr
+            expr = self.simplify(expr)
+
+            static_expr = self._maybe_evaluate_static(expr)
+            if static_expr is not None:
+                return static_expr
+
+            if isinstance(expr, sympy.Eq):
+                self._maybe_guard_eq(expr)
+            concrete_val = self.size_hint(expr)
+
+            # Uncomment this to see what code triggered this guard.
+            # TODO: Save this to the guard representation so you can look
+            # at it later
+            stack = ''.join(traceback.format_stack())
+            self.guards.append((expr, concrete_val, stack))
+            return concrete_val
+        except Exception as e:
+            print(e)
+            print()
```

Comment:
Kill this?

Reply:
Hmm... it is kinda useful....

Reply:
Since otherwise errors in calls to this function often get swallowed

Reply:
Convert it into logging then? Although we kind of need an easy way to only enable logs from some modules and not others.

Reply:
okie dokie I'll just remove the try/catch.